### PR TITLE
Fix handling of removed files in Repo.reset_index

### LINKED ISSUE #1072: Repo.reset_index is broken
Suppose I have a repository with a staged file:
```bash
mkdir test-repo
cd test-repo
git init
git commit --allow-empty --allow-empty-message -m ''
touch file
git add file
```

Then I reset the index using dulwich:
```python
import dulwich.repo
r = dulwich.repo.Repo('test-repo')
r.reset_index()
```

However, the index is not reset. It still contains `file`.

The cause seems to be in `build_index_from_tree`. The documentation claims that the existing index is wiped. However, the function starts with:
```python
index = Index(index_path)
```
This reads in the existing content of the index.

### PR DESCRIPTION
**Pull request by sattlerc**
This fixes #1072.

### CODE REVIEW

**Comment by jelmer** [comment]
Please add a test as well?

**Comment by sattlerc** [comment]
Done.

**Comment by sattlerc** [comment]
The failing test doesn't seem related to this pull request.

File: dulwich/tests/__init__.py [hunk 1]
```diff
@@ -112,41 +112,7 @@ def run_command(self, name, args):
 
 def self_test_suite():
     names = [
-        "archive",
```
Comment (jelmer) [inline dulwich/tests/__init__.py:115]: Uhm, is this just GitHub on mobile acting up or does this remove the list of test modules?

Comment (sattlerc) [reply dulwich/tests/__init__.py:115]: Oops, didn't mean to commit that!

**Comment by sattlerc** [comment]
There's a bunch of trailing whitespace in NEWS that I accidentally deleted in the previous push.

**Pull request merged**